src/adagrams.js

```diff
@@ -1,15 +1,189 @@
+// Helper functions for drawing letters
+const createLetterPoolArray = () => {
+  const letterQuantities = {
+    A: 9,
+    B: 2,
+    C: 2,
+    D: 4,
+    E: 12,
+    F: 2,
+    G: 3,
+    H: 2,
+    I: 9,
+    J: 1,
+    K: 1,
+    L: 4,
+    M: 2,
+    N: 6,
+    O: 8,
+    P: 2,
+    Q: 1,
+    R: 6,
+    S: 4,
+    T: 6,
+    U: 4,
+    V: 2,
+    W: 2,
+    X: 1,
+    Y: 2,
+    Z: 1,
+  };
+  let letterPoolArray = [];
+
+  for (const [letter, quantity] of Object.entries(letterQuantities)) {
+    for (let i = 0; i < quantity; i++) {
+      letterPoolArray.push(letter);
+    }
+  }
+
+  return letterPoolArray;
+};
+
+const getRandomIndex = (arrayLength) => {
+  return Math.floor(Math.random() * arrayLength);
+};
+
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  let letterPoolArray = createLetterPoolArray();
+  let currentHand = [];
+
+  for (let i = 0; i < 10; i++) {
+    const drawnIndex = getRandomIndex(letterPoolArray.length);
+    currentHand.push(letterPoolArray[drawnIndex]);
+    letterPoolArray.splice(drawnIndex, 1);
+  }
+
+  return currentHand;
+};
+
+// Helper functions for counting instances
+const countArray = (array, value) => {
+  let count = 0;
+
+  array.forEach((item) => {
+    if (item === value) {
+      count++;
+    }
+  });
+
+  return count;
+};
+
+const countString = (string, value) => {
+  let count = 0;
+
+  for (let i = 0; i < string.length; i++) {
+    if (string[i] === value) {
+      count++;
+    }
+  }
+
+  return count;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  const upperInput = input.toUpperCase();
+  const letters = new Set(upperInput);
+
+  for (let letter of letters) {
+    const inputCount = countString(upperInput, letter);
+    const handCount = countArray(lettersInHand, letter);
```

Comment:
This approach is O(n^2) with the length of the `letters` list as our `n`, since the outer loop and the `countString` function are O(n) operations. 

There is an O(n) approach where we create frequency maps of the letters in `input` and `lettersInHand`. Then we can iterate over the letters in the `input` frequency map and return false if:
- the letter doesn't exist in `lettersInHand` 
- if the value of the letter is greater in the `input` frequency map than in the `lettersInHand` frequency map

Reply:
I tried doing this, and I've gotten it to pass the regular tests, but the demo tests are failing and I'm not sure why. 
code below:
`export const usesAvailableLetters = (input, lettersInHand) => {
  // const upperInput = input.toUpperCase();
  // const letters = new Set(upperInput);

  // for (let letter of letters) {
  //   const inputCount = countIterable(upperInput, letter);
  //   const handCount = countIterable(lettersInHand, letter);
  //   if (handCount < inputCount) {
  //     return false;
  //   }
  // }

  let handMap = {};
  let inputMap = {};
  for (let letter of lettersInHand) {
    const upperLetter = letter.toUpperCase();
    if (upperLetter in handMap) {
      handMap[upperLetter] += 1;
    } else {
      handMap[upperLetter] = 1;
    }
  }

  for (let letter of input) {
    const upperLetter = letter.toUpperCase();
    if (upperLetter in inputMap) {
      inputMap[upperLetter] += 1;
    } else {
      inputMap[upperLetter] = 1;
    }
  }

  for (let letter in inputMap) {
    if (!letter in handMap) {
      return false;
    } else if (inputMap[letter] !== handMap[letter]) {
      return false;
    }
  }

  return true;
};`

Reply:
I see 2 things that would get this working. We would want `if (!letter in handMap) {` to be `if (!(letter in handMap)) {` since we want the `not` applied to the result of checking in, rather than `letter` itself.

We would also want to change the comparison in `else if (inputMap[letter] !== handMap[letter]) {` to `else if (inputMap[letter] > handMap[letter]) {` since their counts don't need to be exactly equal, we just need the input to not have more than the hand count.

Reply:
Those two changes did the trick, thank you!